### PR TITLE
PIM-7012: add mass associate to product

### PR DESCRIPTION
**Pull request by juliensnz**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added Behats                      | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: features/Context/Page/Base/Form.php [hunk 1]
```diff
@@ -37,7 +38,10 @@ public function __construct($session, $pageFactory, $parameters = [])
                 'Dialog'                          => ['css' => 'div.modal'],
                 'Associations list'               => ['css' => '.associations-list'],
                 'Group selector'                  => ['css' => '.group-selector'],
-                'Association type selector'       => ['css' => '.association-type-selector'],
+                'Association type selector'       => [
```
Comment (SamirBoulil) [inline features/Context/Page/Base/Form.php]: to remove if not used

File: features/Context/catalog/footwear/jobs.yml [hunk 2]
```diff
@@ -280,6 +280,11 @@ jobs:
         alias:     move_to_category
         label:     Mass move to categories
         type:      mass_edit
+    add_association:
```
Comment (SamirBoulil) [inline features/Context/catalog/footwear/jobs.yml]: add in EE fixtures

Comment (juliensnz) [reply features/Context/catalog/footwear/jobs.yml]: done

File: src/Pim/Bundle/InstallerBundle/Resources/fixtures/icecat_demo_dev/jobs.yml [hunk 3]
```diff
@@ -212,11 +212,26 @@ jobs:
...
     add_to_category:
         connector: Akeneo Mass Edit Connector
         alias:     add_to_category
         label:     Mass add to categories
         type:      mass_edit
+    add_association:
+        connector: Akeneo Mass Edit Connector
+        alias:     add_association
+        label:     Mass associate products
+        type:      mass_edit
```
Comment (grena) [inline src/Pim/Bundle/InstallerBundle/Resources/fixtures/icecat_demo_dev/jobs.yml]: I think you just want to add it one ^^ (here you add 3 times the same job)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+                    filePath = item.meta.image.filePath;
+                }
+
+                return MediaUrlGenerator.getMediaShowUrl(filePath, 'thumbnail_small');
+            },
+
+            /**
+             * {@inheritdoc}
+             */
+            labelMethod: function (item) {
```
[the remark this answers is not in the document]
Comment (juliensnz) [reply src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js]: I kept the naming already used on other picker. I agree that it's not perfect, I note it and maybe fix it after the merge

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+            },
+
+            /**
+             * Update the model after dom event triggered
+             *
+             * @param {string} group
+             */
+            setValue: function (values) {
+                const data = this.getFormData();
+                data.actions = [{
```
[the remark this answers is not in the document]
Comment (juliensnz) [reply src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js]: yup, const mean that you cannot reassign the variable

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+                        const matchProductModel = item.match(/^product_model_(.*)$/);
+                        if (matchProductModel) {
+                            productModelIds.push(matchProductModel[1]);
+                        } else {
+                            const matchProduct = item.match(/^product_(.*)$/);
+                            productIds.push(matchProduct[1]);
+                        }
+                    });
+
+                    const assocType = this.getCurrentAssociationType();
```
[the remark this answers is not in the document]
Comment (juliensnz) [reply src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js]: Cannot really happen as this code is executer after a modal confirm and the modal is randered in the render method. And the locale storage is initialized in the render method

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+
+                    const associations = {};
```
[the remark this answers is not in the document]
Comment (juliensnz) [reply src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js]: nope :) Explained above

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/builder.js [hunk 8]
```diff
@@ -20,9 +20,18 @@ define(
             return FormRegistry.getFormExtensions(formMeta).then((extensionsMeta) => {
                 const FormClass = requireContext(formMeta.module);
 
+                if (undefined === FormClass) {
+                    throw new Error(`
+The module "${formMeta.module}" is undefined.
+Most of the time it's because it's not well registered in your requirejs.yml file.
+Here is the documentation to fix this problem
+https://docs.akeneo.com/latest/design_pim/overview.html#register-it`
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/builder.js]: :heart:

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+                this.setValue([]);
+            },
+
+            /**
+             * {@inheritdoc}
+             */
+            render: function () {
+                if (!this.readOnly) {
+                    this.loadAssociationTypes().then((associationTypes) => {
+                        const currentAssociationType = associationTypes.length ? _.first(associationTypes).code : null;
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js]: currentAssociationTypeCode

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+             * @param {string} associationType
+             */
+            setCurrentAssociationType: function (associationType) {
+                sessionStorage.setItem('current_association_type', associationType);
+            },
+
+            /**
+             * @returns {string}
+             */
+            getCurrentAssociationType: function () {
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js]: getCurrentAssociationTypeCode

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+            /**
+             * {@inheritdoc}
+             */
+            render: function () {
+                if (!this.readOnly) {
+                    this.loadAssociationTypes().then((associationTypes) => {
+                        const currentAssociationType = associationTypes.length ? _.first(associationTypes).code : null;
+
+                        if (null === this.getCurrentAssociationType() ||
+                            _.isUndefined(_.findWhere(associationTypes, {code: this.getCurrentAssociationType()}))
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js]: Why not `.filter()` and `undefined ===` ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+                    .always(() => {
+                        loadingMask.remove();
+                    });
+                }
+            },
+
+            /**
+             * {@inheritdoc}
+             */
+            imagePathMethod: function (item) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js]: resolveImagePath() ?

Comment (juliensnz) [reply src/Pim/Bundle/EnrichBundle/Resources/public/js/mass-edit/form/product/associate.js]: used in other modules, I keeep it that way for now

**Pull request merged**